src/core.ts

```diff
@@ -462,126 +462,153 @@ export const noChange = {};
-  setValue(values: any[], startIndex: number): void;
+  /**
+   * Commits the current part value, cause it to actually be written to the DOM.
+   */
+  commit(): void;
 }
 
-export class AttributePart implements MultiPart {
-  instance: TemplateInstance;
+/**
+ * Sets attribute values for AttributeParts, so that the value is only set once
+ * even if there are multiple parts for an attribute.
+ */
+export class AttributeCommitter {
   element: Element;
   name: string;
   strings: string[];
-  size: number;
-  _previousValues: any;
+  parts: AttributePart[];
+  dirty = true;
 
-  constructor(
-      instance: TemplateInstance, element: Element, name: string,
-      strings: string[]) {
-    this.instance = instance;
+  constructor(element: Element, name: string, strings: string[]) {
     this.element = element;
     this.name = name;
     this.strings = strings;
-    this.size = strings.length - 1;
+    this.parts = [];
+    for (let i = 0; i < strings.length - 1; i++) {
+      this.parts[i] = this._createPart();
+    }
+  }
 
-    this._previousValues = [];
+  /**
+   * Creates a single part. Override this to create a differnt type of part.
+   */
+  protected _createPart(): AttributePart {
+    return new AttributePart(this);
   }
 
-  protected _interpolate(values: any[], startIndex: number) {
+  protected _getValue(): any {
     const strings = this.strings;
     const l = strings.length - 1;
     let text = '';
 
     for (let i = 0; i < l; i++) {
       text += strings[i];
-      const v = getValue(this, values[startIndex + i]);
-      if (v && v !== noChange &&
-          (Array.isArray(v) || typeof v !== 'string' && v[Symbol.iterator])) {
-        for (const t of v) {
-          // TODO: we need to recursively call getValue into iterables...
-          text += t;
+      const part = this.parts[i];
+      if (part !== undefined) {
+        const v = part._value;
+        if (v != null &&
+            (Array.isArray(v) || typeof v !== 'string' && v[Symbol.iterator])) {
+          for (const t of v) {
+            text += t;
+          }
+        } else {
+          text += v;
         }
-      } else {
-        text += v;
       }
     }
-    return text + strings[l];
+
+    text += strings[l];
+    return text;
   }
 
-  protected _equalToPreviousValues(values: any[], startIndex: number) {
-    if (this._previousValues.length === 0) {
-      // Always return false for the initial render, otherwise parts with an
-      // undefined value won't get rendered.
-      return false;
+  commit(): void {
+    if (this.dirty) {
+      this.dirty = false;
+      this.element.setAttribute(this.name, this._getValue());
     }
-    for (let i = startIndex; i < startIndex + this.size; i++) {
-      if (this._previousValues[i] !== values[i] ||
-          !_isPrimitiveValue(values[i])) {
-        return false;
+  }
+}
+
+export class AttributePart implements Part {
+  committer: AttributeCommitter;
+  _value: any = undefined;
```

Comment:
Seems wrong that this is `_value` since directives can and do access it. (Same for NodePart).

Reply:
Renamed to `value`